### PR TITLE
Wrap NN operations to TorchTrainer

### PR DESCRIPTION
**Pull request by Eleven1Liu**
**Description** 
1. Put all nn operations to `torch_trainer.py`.
2. Add `--linear` for linear models (see main.py)

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -163,128 +158,30 @@ def main():
...
-            device=device,
-            max_seq_length=config.max_seq_length,
-            batch_size=config.eval_batch_size,
-            data_workers=config.data_workers
-        )
-        trainer.test(model, test_dataloaders=test_loader)
-        if config.save_k_predictions > 0:
-            if not config.predict_out_path:
-                config.predict_out_path = os.path.join(checkpoint_dir, 'predictions.txt')
-            save_predictions(trainer, model, test_loader, config.predict_out_path)
```
Comment (cjlin1) [inline main.py:287]: 284-287 not handled in the new nn training file?? Not nobody calls save_predictions?

File: main.py [hunk 2]
```diff
@@ -148,143 +143,34 @@ def check_config(config):
...
-            model_name=config.model_name, silent=config.silent)
-    else:
-        if config.checkpoint_path:
-            model = Model.load_from_checkpoint(
-                config.checkpoint_path, device=device,
-                model_name=config.model_name, silent=config.silent)
+    # train
+    if not config.eval:
+        if config.linear:
+            # y, x = linear.read_tfidf()
```
Comment (cjlin1) [inline main.py:162]: an issue here is that linear will choose from using raw data or tfidf. So the data loader cannot be specific to nn. See my earlier notes (p2). There is an option to say if raw texts or tfidf would like to be used

Comment (cjlin1) [reply main.py:162]: had some discussion with my students. Loader for NN is way more complicated than linear needs. So they will use different loaders. Thus your setting now is basically ok.

File: main.py [hunk 3]
```diff
@@ -148,143 +143,34 @@ def check_config(config):
...
-                         num_sanity_val_steps=0,
-                         gpus=0 if config.cpu else 1,
-                         progress_bar_refresh_rate=0 if config.silent else 1,
-                         max_epochs=config.epochs,
-                         callbacks=[checkpoint_callback, earlystopping_callback])
 
-    # Dump config to log
-    log_path = os.path.join(checkpoint_dir, 'logs.json')
-    dump_log(log_path, config=config)
+    if not config.linear:
```
Comment (cjlin1) [inline main.py:156]: This place can now be
if config.linear:
# load raw texts and generate tfidf, or load tfidf
else:
trainer = ...

File: main.py (hunk 2 above)
Comment (cjlin1) [inline main.py:162]: this line can be removed as data loading for linear occurs in the earlier place

Comment (Eleven1Liu) [reply main.py:162]: Sure. got it!

**Review by cjlin1 — APPROVED** [review]

**Pull request merged**